### ISSUE TITLE
Option to save and retrieve Cookies as dicts instead of strings

### ISSUE THREAD

**Issue by wlerin** [ping/instagram_private_api#4]
### Which client are you using?
- [x] app (``instagram_private_api/``)
- [ ] web (``instagram_web_api/``)

### Purpose of your issue?
- [x] Other

---

### Proposed rework of ClientCookieJar

I'm used to using Requests' syntax for options and cookies, and that library allows you to save and restore cookies in dict format (and thus, as JSON). By default, ClientCookieJar uses `pickle.dumps`, which returns a `bytes` object that cannot be directly stored in JSON. 

I *could* encode it as base64 and then decode it as utf-8, but the resulting string is not human readable, and I still have to reverse these two steps to restore the cookies. Or I could dump it directly to a file in 'wb' mode, but all my config files are JSON.

Anyway, long story short, I have written a modified version of ClientCookieJar that will accept either a `dict` or a `bytes` object as its `cookie_repr` argument, and remember which one was used when it comes time to dump its cookies. 

The format can also be set after CookieJar creation by setting the `format` property to `"dict"` or `"json"` for dict output, or `"pickle"` or `"bytes"` or `"str"` for bytes output.

[Here are the proposed changes](https://github.com/ping/instagram_private_api/compare/master...social48:cookies).

**Comment by ping** [comment]
Doesn't pickle.dumps return a string? https://docs.python.org/2/library/pickle.html#pickle.dumps

I haven't had any problems saving the pickled string to a json file. I even wrote an example here https://github.com/ping/instagram_private_api/blob/master/examples/savesettings_logincallback.py based off how I save my settings. Does that not work for you?

**Comment by wlerin** [comment]
`pickle.dumps` _always_ returns bytes, the difference is that in Python 2, `bytes` is also `str` and gets automagically handled by the interpreter. In [Python 3](https://docs.python.org/3/library/pickle.html#pickle.dumps) with your example you get this instead:

> TypeError: Object of type 'bytes' is not JSON serializable

**Comment by ping** [comment]
Thanks for the catch. Please submit your changes in a PR and I'll get it merged.

**Comment by wlerin** [comment]
It's actually possible to solve my original issue by just adding a .decode("utf-8") [here](https://github.com/ping/instagram_private_api/blob/master/instagram_private_api/http.py#L24).

I still think letting cookies be saved and restored as dicts is a useful feature, but it's not absolutely necessary to solve this problem.

**Comment by ping** [comment]
Does that work for you?

I gave it a quick test but got
``UnicodeDecodeError: 'utf-8' codec can't decode byte 0x80 in position 0: invalid start byte``

With Py3.5.1

**Comment by wlerin** [comment]
Ah. I forgot one detail. This works in 2.7 and 3.6:
```python
pickle.dumps({"test": "test"}, protocol=1).decode('utf-8')
```
However I still need to check whether it works for Cookies instead of simple dicts.

**Comment by wlerin** [comment]
I'm going to keep refining and testing a CookieJar that accepts both dicts and `str|bytes`, but until then I'm closing this PR as the new changes are sufficient for saving to JSON.
